### PR TITLE
IKC-242 SSO

### PR DESCRIPTION
**Pull request by pbelke**

### CODE REVIEW

File: kouncil-backend/src/main/java/com/consdata/kouncil/config/security/ad/ActiveDirectoryWebSecurityConfig.java [hunk 1]
```diff
@@ -27,20 +33,39 @@ public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
...
                 })
                 .and()
                 .authorizeRequests()
-                .antMatchers("/api/info/version").permitAll()
-                .antMatchers("/api/firstTimeLogin").permitAll()
-                .antMatchers("/api/login").permitAll()
-                .antMatchers("/api/**").authenticated()
-                .antMatchers("/**").permitAll();
+                .antMatchers("/api/info/version", "/api/login", "/**").permitAll()
+                .anyRequest().authenticated();
```
Comment (Pefes) [inline kouncil-backend/src/main/java/com/consdata/kouncil/config/security/ad/ActiveDirectoryWebSecurityConfig.java:45]: Would it work? Every url would be matched by "/**"

File: kouncil-frontend/libs/common-login/src/lib/login/common-login.component.ts [hunk 2]
```diff
@@ -29,14 +31,37 @@ import {User} from "./user";
         </button>
       </form>
     </div>
+
+    <div class="sso-container" *ngIf="availableProviders && availableProviders.length >0">
```
Comment (dmejer) [inline kouncil-frontend/libs/common-login/src/lib/login/common-login.component.ts:35]: this template should have own component

File: kouncil-frontend/apps/kouncil/src/app/login/login.component.ts [hunk 3]
```diff
@@ -28,12 +29,14 @@ import {environment} from '../../environments/environment';
...
 
   firstTimeLogin: boolean = false;
+  providers: Array<string>;
   public backend: Backend = environment.backend;
 
   constructor(private service: AuthService, private router: Router) {
   }
 
   ngAfterViewInit(): void {
+    this.service.ssoProviders$().subscribe(providers => this.providers = providers);
```
Comment (dmejer) [inline kouncil-frontend/apps/kouncil/src/app/login/login.component.ts:39]: this kind of logic should be in OnInit instead afterViewInit

File: kouncil-backend/src/main/java/com/consdata/kouncil/config/security/inmemory/InMemoryWebSecurityConfig.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+@EnableWebSecurity
+@Configuration
+@Slf4j
+@RequiredArgsConstructor
+@ConditionalOnProperty(prefix = "kouncil.auth", name = "active-provider", havingValue = "inmemory")
+public class InMemoryWebSecurityConfig {
+
+    @Bean
+    public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
```
Comment (dmejer) [inline kouncil-backend/src/main/java/com/consdata/kouncil/config/security/inmemory/InMemoryWebSecurityConfig.java:35]: Can we maintance only one security config? LdapWebSecurityConfig vs  InMemoryWebSecurityConfig?
Some should test this independently  ;)

File: kouncil-backend/src/main/java/com/consdata/kouncil/config/security/inmemory/InMemoryWebSecurityConfig.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+                    configuration.setAllowedMethods(List.of("*"));
+                    configuration.setAllowedHeaders(List.of("*"));
+
+                    UrlBasedCorsConfigurationSource source = new UrlBasedCorsConfigurationSource();
+                    source.registerCorsConfiguration("/**", configuration);
+                    return configuration;
+                })
+                .and()
+                .authorizeRequests()
+                .antMatchers("/api/info/version", "/api/firstTimeLogin", "/api/login", "/**").permitAll()
```
Comment (Pefes) [inline kouncil-backend/src/main/java/com/consdata/kouncil/config/security/inmemory/InMemoryWebSecurityConfig.java:50]: Same doubts as in ActiveDirectoryWebSecurityConfig

Comment (dmejer) [inline kouncil-backend/src/main/java/com/consdata/kouncil/config/security/inmemory/InMemoryWebSecurityConfig.java:50]: "/**" do you whant to permit all without auth all endpoints?

File: kouncil-backend/src/main/java/com/consdata/kouncil/config/security/sso/TokenStore.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+import org.springframework.security.core.Authentication;
+import org.springframework.stereotype.Component;
+
+@Component
+public class TokenStore {
+
+    private final Map<String, Authentication> cache = new HashMap<>();
+
+    public String generateToken(Authentication authentication) {
+        String token = UUID.randomUUID().toString();
```
Comment (dmejer) [inline kouncil-backend/src/main/java/com/consdata/kouncil/config/security/sso/TokenStore.java:15]: what about jwt token?

File: kouncil-backend/src/main/java/com/consdata/kouncil/config/security/inmemory/InMemoryWebSecurityConfig.java (hunk 4 above)
Comment (dmejer) [reply kouncil-backend/src/main/java/com/consdata/kouncil/config/security/inmemory/InMemoryWebSecurityConfig.java:35]: I will try this tomorrow ;)

File: kouncil-frontend/apps/kouncil/src/app/routing/routing.module.ts [hunk 7]
```diff
@@ -93,13 +94,13 @@ const routes: Routes = [
...
     ]
-  }
+  },
+  {path: 'oauth', component: OAuthRedirectComponent}
 ];
 
 @NgModule({
   imports: [
     RouterModule.forRoot(routes, {
-      useHash: true,
```
Comment (dmejer) [inline kouncil-frontend/apps/kouncil/src/app/routing/routing.module.ts:102]: why not?

**Review by dmejer — CHANGES_REQUESTED** [review]

File: kouncil-backend/src/main/java/com/consdata/kouncil/config/security/sso/TokenStore.java (hunk 6 above)
Comment (pbelke) [reply kouncil-backend/src/main/java/com/consdata/kouncil/config/security/sso/TokenStore.java:15]: Added todo task

File: kouncil-frontend/apps/kouncil/src/app/routing/routing.module.ts (hunk 7 above)
Comment (pbelke) [reply kouncil-frontend/apps/kouncil/src/app/routing/routing.module.ts:102]: If you want to authenticate with github sso you have to pass callback url. If this url contains #, everything from this sign will be placed at the end of the response url

File: kouncil-backend/src/main/java/com/consdata/kouncil/config/security/inmemory/InMemoryWebSecurityConfig.java (hunk 4 above)
Comment (dmejer) [reply kouncil-backend/src/main/java/com/consdata/kouncil/config/security/inmemory/InMemoryWebSecurityConfig.java:35]: You said you would valid token in next MR. Im ok with that

**Review by dmejer — APPROVED** [review]

**Pull request merged**